core/src/main/scala/org/apache/spark/sql/hive/SnappyStoreHiveCatalog.scala

```diff
@@ -615,6 +622,44 @@ class SnappyStoreHiveCatalog(context: SnappyContext)
     }
   }
 
+  def alterTableToAddIndexProp(inTable: QualifiedTableName, indexName: String): Unit = {
+    alterTableLock.synchronized {
+      val hiveTable = inTable.getTable(client)
+      var indexes = ""
+      try {
+        indexes = hiveTable.serdeProperties(ExternalStoreUtils.INDEX_NAME) + ","
+      } catch {
+        case e: NoSuchElementException =>
+      }
+      client.alterTable(
+        hiveTable.copy(serdeProperties = hiveTable.serdeProperties +
+          (ExternalStoreUtils.INDEX_NAME -> (indexes + indexName)))
+      )
+    }
```

Comment:
note the code in registerDataSourceTable/unregisterDataSourceTable. We catch HiveException and then retry with a new client. This is to handle "stale" connections which are particularly the case in our unit tests with repeatedly start/shutdown the database (there are other cases too). Add similar handling here.